python/sgl_jax/srt/models/qwen2_moe.py

```diff
@@ -0,0 +1,605 @@
+                    target_path=f"{target_prefix}.self_attn.v_proj.bias",
+                    sharding=(None,),
+                    transpose=False,
+                    head_dim_padding=True,
+                    kv_head_padding=True,
+                ),
+                f"{prefix}.self_attn.o_proj.bias": WeightMapping(
+                    target_path=f"{target_prefix}.self_attn.o_proj.bias",
+                    sharding=(None,),
+                    transpose=False,
+                ),
+            }
+            mappings.update(bias_mappings)
+
+        # MoE mappings for expert layers
+        mappings[f"{prefix}.mlp.gate.weight"] = WeightMapping(
+            target_path=f"{target_prefix}.moe_gate.kernel",
+            sharding=(None, None),
+            transpose=True,
+        )
+
+        # Optional shared expert weight mapping (singular in source naming)
+        if (
+            getattr(self.config, "shared_expert_intermediate_size", 0)
+            and getattr(self.config, "shared_expert_intermediate_size", 0) > 0
+        ):
+            shared_expert_mappings = {
+                f"{prefix}.mlp.shared_expert.gate_proj.weight": WeightMapping(
+                    target_path=f"{target_prefix}.shared_experts.gate_proj.weight",
+                    sharding=(None, "tensor"),
+                    transpose=True,
+                ),
+                f"{prefix}.mlp.shared_expert.up_proj.weight": WeightMapping(
+                    target_path=f"{target_prefix}.shared_experts.up_proj.weight",
+                    sharding=(None, "tensor"),
+                    transpose=True,
+                ),
+                f"{prefix}.mlp.shared_expert.down_proj.weight": WeightMapping(
+                    target_path=f"{target_prefix}.shared_experts.down_proj.weight",
+                    sharding=("tensor", None),
+                    transpose=True,
+                ),
+                f"{prefix}.mlp.shared_expert_gate.weight": WeightMapping(
+                    target_path=f"{target_prefix}.shared_expert_gate.weight",
+                    sharding=(None, None),
+                    transpose=True,
+                ),
+            }
+            mappings.update(shared_expert_mappings)
+
+        num_experts = getattr(self.config, "num_experts", 8)
+        for expert_type in ["gate_proj", "up_proj", "down_proj"]:
+            target_name = {
+                "gate_proj": "wi_0",
+                "up_proj": "wi_1",
+                "down_proj": "wo",
+            }[expert_type]
+            expert_keys = [
+                f"{prefix}.mlp.experts.{i}.{expert_type}.weight" for i in range(num_experts)
+            ]
+
+            mappings[f"__MOE_EXPERTS__{prefix}.mlp.{target_name}"] = WeightMapping(
+                target_path=[f"{target_prefix}.mlp.{target_name}"] + expert_keys,
+                sharding=(("data", "tensor"), None, None),
+                transpose=True,
+            )
```

Comment:
Using Expert Parallelism and Tensor Parallelism involves different sharding approaches. Refer to qwen3 moe.